### ISSUE TITLE
Several questions about albumentations

### ISSUE THREAD

**Issue by RandomGuy34** [albumentations-team/albumentations#650]
Hi everyone,

I decided to switch to the `albumentations` library for practical purposes, and after few days of trying it I have some questions related to its good use. Note that I am a beginner in the deep learning field, thus I might do dumb mistakes. Do not hesitate to tell me if so.

First of all, I have to tell you more about my CNN model : it is some kind of autoencoder, but the expected output images are different from my input images. Furthermore, I have thousands of samples, thus I need to work with an array of images paths rather than with an array containing all my huge dataset.

For these two reasons I cannot use the Keras `ImageDataGenerator` class, and so that's why I chose to work with `albumentations`, a more flexible library. You can see below my source code (I tried to be the most concise possible, thus I did not detailed my CNN code).

However, I encountered 3 issues with this data augmentation tool : 

1. First of all, it seems that _[Pixel-level transforms](https://github.com/albumentations-team/albumentations#pixel-level-transforms)_ produce black images (whereas _[Spatial-level transforms](https://github.com/albumentations-team/albumentations#spatial-level-transforms)_ do not) on images loaded with the `load_img` and `img_to_array` functions from `tf.keras.preprocessing.image`. It took me several hours to figure out the necessity of `cv2.imread` from OpenCV, but I still haven't understood why this wouldn't work, and I'd be interested by any explanation :)

2. Why does it take so much time? A training with `albumentations` (and even with only the basic `ToFloat(max_value=255)` function) takes ~3 times longer than a training without data augmentation (for which I used a sligthly different version of `batch_generator_with_paths`, without `self.augment` and using `cv2.normalize(img, None, 0, 1, cv2.NORM_MINMAX, cv2.CV_32F)` to rescale the pixels between [0,1]).

3. Why is my training loss so different ? You can find below an example of the loss I obtained without data augmentation on a small dataset (~540 train images, with `nb_epochs=5`). 
![loss without data aug](https://user-images.githubusercontent.com/66247190/87047484-de131a00-c1fa-11ea-87ec-fc20523a30ea.png)

 I had quite the same curve and the same loss values with `ImageDataGenerator`, but on the contrary `albumentations` gives random curves (really different from one training to another) and really high loss values :

![loss with albu](https://user-images.githubusercontent.com/66247190/87047502-e53a2800-c1fa-11ea-9da8-0c97dd4a215f.png)
![loss with albu 2](https://user-images.githubusercontent.com/66247190/87047508-e8351880-c1fa-11ea-94c7-f419ed95cf5e.png)

I truly don't understand those graphs and the loss values, I thought the loss of an autoencoder would give the average (for all the training images) MSE on each dimension of the image (as described [here](https://datascience.stackexchange.com/questions/15568/understanding-autoencoder-loss-function)).

Any help would be greatly appreciated !

### Source code :

    # Colab Notebook
    ###########  Defining the constants  ###########
    
    train_list = ... # numpy array, containing the paths to the training images.
    labels_train_list = ... # numpy array, containing the paths to the labels images. labels_train_list[i] gives the path to the label image of train_list[i].
    test_list = ... # same for the test list.
    labels_test_list = ...
    batch_size = ...
    nb_epochs = ...
    
    
    ###########  Creating the data augmentation functions  ###########  
    
    data_generator_with_aug = Compose([HorizontalFlip(p = 0.5),
                                       RandomContrast(limit = 0.1, p = 0.2),
                                       RandomGamma(gamma_limit = (80, 120), p = 0.2),
                                       RandomBrightness(limit = 0.1, p = 0.2),
                                       ToFloat(max_value = 255)
                                       ])
    data_generator_no_aug = ToFloat(max_value = 255)
    
    
    ###########  Creating a batch generator (subclass of Keras Sequence class) working with the images paths  ###########  
    
    class batch_generator_with_paths(Sequence)  :  # from tensorflow.python.keras.utils.data_utils import Sequence
      
        def __init__(self, x_set, y_set, batch_size, augmentations, shuffle = True) :
            self.x, self.y = x_set, y_set
            self.batch_size = batch_size
            self.augment = augmentations
            self.shuffle = shuffle
            self.indexes = np.arange(len(self.x))

        def __len__(self) :
            return int(np.ceil(len(self.x) / float(self.batch_size)))

        def __getitem__(self, idx) :
            inds = self.indexes[idx * self.batch_size:(idx + 1) * self.batch_size]
            batch_x = self.x[inds]
            batch_y = self.y[inds]
            return np.array([self.augment(image = cv2.cvtColor(cv2.imread(filename), cv2.COLOR_RGB2BGR) )["image"] for filename in batch_x
                             ]), np.array([cv2.cvtColor(cv2.imread(filename), cv2.COLOR_RGB2BGR) for filename in batch_y])
      
        def on_epoch_end(self)  :
            self.indexes = np.arange(len(self.x))
            if self.shuffle :
              np.random.shuffle(self.indexes)
    
    
    ###########  Creating the train and test generators  ###########
    
    train_generator = batch_generator_with_paths(train_list, labels_train_list, batch_size = batch_size, augmentations = data_generator_with_aug, shuffle=True)
    test_generator = batch_generator_with_paths(test_list, labels_test_list, batch_size = 1, augmentations = data_generator_no_aug)
    
    
    ###########  Creating and training the model (I have a ConvAutoencoder class defining my CNN autoencoder)  ###########

    (encoder, decoder, autoencoder) = ConvAutoencoder.build(width, height, depth) # width, height and depth being predefined constants.
    opt = Adam(lr=1e-3) # from tensorflow.keras.optimizers import Adam
    autoencoder.compile(loss="mse", optimizer=opt)

    H = autoencoder.fit(train_generator,
                        steps_per_epoch = len(train_list) // batch_size,
                        epochs = nb_epochs,
			    verbose = 1
			    )
